sdk/python/lib/pulumi/log.py

```diff
@@ -65,14 +73,23 @@ def error(msg, resource=None, stream_id=None):
     else:
         print("error: " + msg, file=sys.stderr)
 
-def _log(engine, severity, message, resource, stream_id):
-    if resource is not None:
-        urn = resource.urn
-    else:
-        urn = ""
 
+def _log(engine, severity, message, resource, stream_id):
     if stream_id is None:
         stream_id = 0
 
-    req = engine_pb2.LogRequest(severity=severity, message=message, urn=urn, streamId=stream_id)
-    engine.Log(req)
+    # If we can log synchronously, do so. The worst thing we can do with a log message is exit
+    # before we have the chance to send the message.
+    #
+    # We can log synchronously as long as we haven't been given a resource to attach to. If we have,
+    # we have to asynchronously resolve the URN first.
+    async def do_log():
```

Comment:
IIRC, we have special logic in the Node.js SDK to avoid exiting before outbound messages have drained, at least in the normal exit case. Am I misremembering and/or are we covered here?

Reply:
We do, but that doesn't work well here. We require an active event loop in order to do anything asynchronously and we only boot the event loop right before executing a user's program. We have had bugs in the past with Python that we were not able to log early or late errors because we've already shut down the runtime.

I think this is sufficient for now and we should only try to change it if we find this doesn't work.